### PR TITLE
ARROW-11362:[Rust][DataFusion] Use iterator APIs in to_array_of_size to improve performance

### PR DESCRIPTION
**Pull request by Dandandan**
This function `to_array_of_size` is about 8.3% of total instructions in the db-benchmark (aggregation) queries.

This uses the PR https://github.com/apache/arrow/pull/9293

The case of converting an int32 to an array improved by ~5x according to the microbenchmark:

```
to_array_of_size 100000 time:   [55.501 us 55.627 us 55.809 us]                                    
                        change: [-82.457% -82.384% -82.299%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

And on TCPH query 1 (SF=1, 16 partitions).

PR:

```
Query 1 iteration 0 took 90.8 ms
Query 1 iteration 1 took 106.6 ms
Query 1 iteration 2 took 101.1 ms
Query 1 iteration 3 took 101.5 ms
Query 1 iteration 4 took 96.9 ms
Query 1 iteration 5 took 100.3 ms
Query 1 iteration 6 took 99.6 ms
Query 1 iteration 7 took 100.4 ms
Query 1 iteration 8 took 104.2 ms
Query 1 iteration 9 took 100.3 ms
Query 1 avg time: 100.18 ms
```
Master:
```
Query 1 iteration 0 took 121.1 ms
Query 1 iteration 1 took 123.4 ms
Query 1 iteration 2 took 121.0 ms
Query 1 iteration 3 took 121.0 ms
Query 1 iteration 4 took 123.0 ms
Query 1 iteration 5 took 121.7 ms
Query 1 iteration 6 took 121.7 ms
Query 1 iteration 7 took 120.2 ms
Query 1 iteration 8 took 119.7 ms
Query 1 iteration 9 took 121.4 ms
Query 1 avg time: 121.43 ms
```

### CODE REVIEW

**Review by jorgecarleitao — APPROVED** [review]
LGTM. 

There is a `fmt` issue somewhere.

**Comment by jorgecarleitao** [comment]
I noticed that this is also being used in `hash_aggregate::create_batch_from_map`.

**Comment by Dandandan** [comment]
@jorgecarleitao yes, but those are on `GroupByScalar` instead of `ScalarValue`. I think it might clean up some code by having just one  `ScalarValue` enum defined and reusing functions like this?

**Comment by Dandandan** [comment]
CI failures don't seem to be related to the changes.

File: rust/datafusion/src/scalar.rs [hunk 1]
```diff
@@ -205,28 +205,104 @@ impl ScalarValue {
...
+                        .take(size)
+                        .collect::<TimestampNanosecondArray>(),
+                ),
+            },
+            ScalarValue::Utf8(e) => match e {
+                Some(value) => {
+                    Arc::new(StringArray::from_iter_values(repeat(value).take(size)))
+                }
+                None => {
+                    Arc::new(repeat(e.as_deref()).take(size).collect::<StringArray>())
```
Comment (jorgecarleitao) [inline rust/datafusion/src/scalar.rs:293]: any reason to use `e.as_deref` instead of `None`? Just curious.

Comment (Dandandan) [reply rust/datafusion/src/scalar.rs:293]: I couldn't get the `None` compiling here.

Comment (Dandandan) [reply rust/datafusion/src/scalar.rs]: Updated.

**Comment by alamb** [comment]
The clippy failures were fixed in https://github.com/apache/arrow/pull/9314 - a quick rebase can probably get this PR green

**Comment by Dandandan** [comment]
PR is updated against master @jorgecarleitao @alamb

**Review by alamb — APPROVED** [review]
I reviewed this  PR and it looks good to me.

**Comment by jorgecarleitao** [comment]
@Dandandan there is a missing license in `rust/datafusion/benches/scalar.rs`.

**Comment by wesm** [comment]
Whoa. This PR needed to be rebased

**Comment by jorgecarleitao** [comment]
I am really sorry, @wesm . What were the consequences? 

From what I see, the main problem is that the list of co-authors is rather large on the commit message.

**Comment by Dandandan** [comment]
I think it was still the recent rebase after 3.0 on master (which I think introduced some weirdness on the master branch?) causing this, I missed that it still had those commits in here.

**Comment by alamb** [comment]
All the more reason not to be rebasing master going forward I would say

**Pull request closed without merging**